### ISSUE TITLE
ERROR: 0:407: 'limitations' : Non-constant-index-expression

### ISSUE THREAD

**Issue by intractabilis** [vsg-dev/VulkanSceneGraph#673]
**Describe the bug**
I am trying to use `vsgviewer` from `vsgExamples` and it prints out a long list of some source code, plus above error message, and then exits with a C++ exception
```
/home/user/dev/sandbox/vsg/build/debug/examples/app/vsgviewer/vsgviewer /home/user/libs-dev/glTF-Sample-Models/1.0/Box/glTF-Embedded/Box.gltf 
[Exception] - Error: vsg::ShaderModule::create(...) failed to create shader module. result = -13
```

**To Reproduce**

```
vsgviewer glTF-Sample-Models/1.0/Box/glTF-Embedded/Box.gltf
```
`Box.gltf` file is available [here](https://github.com/KhronosGroup/glTF-Sample-Models).

**Expected behavior**

I don't know. What it is supposed to do without a C++ exception?  I didn't use it before, there is no documentation, so I assume the expected behavior is to show me the model?

**Desktop (please complete the following information):**
 - OS: Manjaro Linux 22.0.0

**Additional context**

I tried finding information about the above error. The only thing I could find is that `TBuiltInResource::limits.generalUniformIndexing` has to be set to `true`. I checked `src/vsg/utils/ShaderCompiler.cpp` with a debugger, it is true.

**Comment by intractabilis** [comment]
Ok, I found where the problem was. `vsg_glslangConfig.cmake` messed up the libraries that had to be compiled. I removed it from `vsgConfig.cmake` and configured `glslang` dependencies manually. I run everything from a separate from the system prefix, located in my home directory. For some inexplicable reason, `vsg_glslangConfig.cmake` picked the libraries from `/usr/lib` instead.

**Comment by robertosfield** [comment]
The intention is that the VSG's cmake build system will find and set up the config files based on the available libraries.  For some reason this looks like it's not worked properly - one should need to delete vsgConfig.cmake and manually set things.

What version of Vulkan and glslang do you have on your system and where are they located?

Could you do a clean build of the VSG without modifying anything and post the CMakeCache.txt files that it generates.  This might help us figure our what is going wrong.

**Comment by intractabilis** [comment]
There is no need to delete `vsgConfig.cmake`, it works. I just deleted the line from it that includes `vsg_glslangConfig.cmake`.

> What version of Vulkan and glslang do you have on your system and where are they located?

Vulkan is lokated in `/usr` prefix, versions are:
```
$ yay -Qs vulkan | grep -A 1 'vulkan-devel'
local/spirv-tools 2022.1-1 (vulkan-devel)
    API and commands for processing SPIR-V modules
--
local/vulkan-extra-layers 1.3.231.1-1 (vulkan-devel)
    Extra layers for Vulkan development
local/vulkan-extra-tools 1.3.231.1-1 (vulkan-devel)
    Vulkan lunarg tools
local/vulkan-headers 1:1.3.235-1 (vulkan-devel)
    Vulkan header files
local/vulkan-html-docs 1:1.3.235-1 (vulkan-devel)
    Vulkan html documentation
--
local/vulkan-tools 1.3.226-1 (vulkan-devel)
    Vulkan Utilities and Tools
local/vulkan-validation-layers 1.3.231.1-1 (vulkan-devel)
    Vulkan Validation Layers
```
glslang I compile myself from `sdk-1.3.236` branch, and it is located in my personal prefix.

> Could you do a clean build of the VSG without modifying anything and post the CMakeCache.txt files that it generates. This might help us figure our what is going wrong.

I am not certain that I understand the task. `vsg_glslangConfig.cmake` is not involved into building of VSG. This is a configuration file that VSG _installs_, and it only affects the code that _uses_ VSG. Do you still want me to build VSG?

**Comment by robertosfield** [comment]
You shouldn't need to modify vsgConfig.cmake, the key to resolving the issues will be figuring out why vsg_glslangConfig.cmake is failing to provide the correct libs.

Where did you install Vulkan from?  Did this install also pull in a glslang library and headers?

Why did you need to build glslang?

How do you tell cmake where the Vulcan and glslang lis and headers are?  Are these automatically detected or do you manually set them in cmake via command line or ccmake?

The CMakeCache.txt that is generated when you build the VulkanSceneGraph includes all the paths, so should help us figure out what it's trying to link against and where they are located.  Some of this information will also be used in generating the vsgConfig,cmake and  vsg_glslangConfig.cmake files.

My guess is something in your set up/usage case has broken the VSG's CMake finding of glslang.  Once we know what it is about this usage case that is causing problems we can try and come up with CMake fix for this, or document how to invoke cmake to avoid problems for this type of usage.

**Comment by intractabilis** [comment]
> Where did you install Vulkan from?
```
sudo pacman -S --needed vulkan-devel
```

> Did this install also pull in a glslang library and headers?

Yes.

> How do you tell cmake where the Vulcan and glslang lis and headers are? Are these automatically detected or do you manually set them in cmake via command line or ccmake?

`find_package`. Yes, manually specifying `find_package` picks up the correct libraries from the correct prefix.

> The CMakeCache.txt that is generated when you build the VulkanSceneGraph includes all the paths, so should help us figure out what it's trying to link against and where they are located.

And you want me to do it using the system prefix, and the system toolset? Or using my development setup that gave me an error?

> Some of this information will also be used in generating the vsgConfig,cmake and vsg_glslangConfig.cmake files.

This is incorrect. You don't generate `vsg_glslangConfig.cmake`.

> My guess is something in your set up/usage case has broken the VSG's CMake finding of glslang.

I think the reason is that you manually search for libraries instead of using `find_package` inside `vsg_glslangConfig.cmake`.

**Comment by robertosfield** [comment]
I'm trying to figure out what is going wrong so I need the CMakeCache.txt file for the problem build.  

Ideally I want a way of recreating the problem on my own system so if I can come up with a possible fix I'll be able to test it, this is why I'm asking for specifics of what you are doing,

You didn't answer why you need to do a local build of glslang.  This is useful for me to know so that I can take account of the usage case better if needed.

FYI, vsg_glslangConfig.cmake is needed because glslang comes in some many different versions and configurations, only very recent ones have find_package() support so for the older versions we have to find things manually as CMake itself doesn't provide it.

**Comment by intractabilis** [comment]
> I'm trying to figure out what is going wrong so I need the CMakeCache.txt file for the problem build.

Do you want me to generate it with the system prefix or with the one that gave me an error? I am asking because I think the latter seems more helpful, but you asked for a clean build in the beginning, so I want to confirm which one you want.

> You didn't answer why you need to do a local build of glslang.

I honestly don't remember how it ended up in my prefix. 😀 I have a script that builds a toolset and a script to build libraries that I need with this toolset. I guess at some point I wanted glslang to be built with my toolset, I don't remember why. Maybe as some dependency.

**Comment by robertosfield** [comment]
Back from hols and working towards a 1.0.3 maintenance release.  It would be good to get to the bottom of the issue you've seen for this release.

I don't know quite how to ask for the CMakeCache.text from the problem build, I keep asking for it and you keep replying without the CMakeCache,txt as this there is some remaining clarification needed.  It really is very simple, I'll restate again in case this somehow makes more sense this time. 

I want the CMakeCache.txt file for the build where the errors in the build occurred - the scenario that started you thread with.  I want to figure what went wrong and if possible recreate it on my system so I can then experiment with possible fixes or documented guidance on how to avoid this issue.

I can't do anything more to resolve the problem without the CMakeCache.txt and/or a means of create the problem.  If this can't be provided I'll just have to close this issue and chalk it up as a random error that user CMake configuration error somewhere on their system.

**Comment by intractabilis** [comment]
I hope you had good holidays. Thank you for working on resolving this issue, and I apologize for miscommunications.

> Are you building the VSG and vsgExamples with different library and include paths?

No, I am not doing anything stupid.

Good guesses, you are almost right. I checked how VSG was built. First, it also ignores the prefix, when it searches for `libglslang.so`, it finds it in `/usr/lib` instead of in my prefix. Here:
```
$ mkdir VulkanSceneGraph-build
$ cd VulkanSceneGraph-build
$ cmake -Wno-dev -LA -G Ninja -DCMAKE_INSTALL_PREFIX=${HOME}/gcc-env -DCMAKE_CXX_COMPILER=g++ -DCMAKE_BUILD_TYPE=Debug ../VulkanSceneGraph
$ grep -RHn libglslang.so .
./CMakeCache.txt:341:glslang_LIBRARY:FILEPATH=/usr/lib/libglslang.so
```
I guess you are not using `find_package` here either because
```
$ find ${HOME}/gcc-env/lib -name libglslang.so
/home/me/gcc-env/lib/libglslang.so
$ ls -AlFd ${HOME}/gcc-env/lib/cmake/glslang
drwxr-xr-x 1 user user 196 янв  2 22:06 /home/user/gcc-env/lib/cmake/glslang/
```
However, by default you are building a static library, so in fact `libglslang.so` is not being used in the build process of VSG at all.

What happens next is you install `vsg_glslangConfig.cmake` to `${HOME}/gcc-env/lib/cmake`, which also ignores the prefix, finds the system libraries, instead of from the prefix, and links with them:
```
$ mkdir vsgExamples-build
$ cd vsgExamples-build
$ cmake -Wno-dev -LA -G Ninja -DCMAKE_INSTALL_PREFIX=${HOME}/gcc-env -DCMAKE_CXX_COMPILER=g++ -DCMAKE_BUILD_TYPE=Debug ../vsgExamples
$ ninja -v | grep vsgviewer
[93/118] /home/me/gcc-env/bin/g++   -march=native -mtune=native -g -Wall -Wparentheses -Wno-long-long -Wno-import -Wreturn-type -Wmissing-braces -Wunknown-pragmas -Wmaybe-uninitialized -Wshadow -Wunused -Wno-misleading-indentation -Wextra -std=gnu++17 -MD -MT examples/app/vsgviewer/CMakeFiles/vsgviewer.dir/vsgviewer.cpp.o -MF examples/app/vsgviewer/CMakeFiles/vsgviewer.dir/vsgviewer.cpp.o.d -o examples/app/vsgviewer/CMakeFiles/vsgviewer.dir/vsgviewer.cpp.o -c /home/me/dev/sandbox/vsgExamples/examples/app/vsgviewer/vsgviewer.cpp
[108/118] : && /home/me/gcc-env/bin/g++ -march=native -mtune=native -g  examples/app/vsgviewer/CMakeFiles/vsgviewer.dir/vsgviewer.cpp.o -o bin/vsgviewer  /home/me/gcc-env/lib/libvsgd.a  /lib/libvulkan.so  /usr/lib/libMachineIndependent.a  /usr/lib/libOSDependent.a  /usr/lib/libOGLCompiler.a  /usr/lib/libSPIRV.so  /usr/lib/libGenericCodeGen.a  /usr/lib/libSPIRV-Tools-opt.so  /usr/lib/libSPIRV-Tools.so  /usr/lib/libxcb.so && :
```
Finally, when the built binaries are being executed, they of course load the libraries from the prefix, instead of what you link them with:
```
$ ldd bin/vsgviewer
        linux-vdso.so.1 (0x00007fffcd9a2000)
        libvulkan.so.1 => /usr/lib/libvulkan.so.1 (0x00007f2d73a4e000)
        libSPIRV.so => /home/me/gcc-env/lib/libSPIRV.so (0x00007f2d72a00000)
        libSPIRV-Tools-opt.so => /usr/lib/libSPIRV-Tools-opt.so (0x00007f2d72600000)
        libSPIRV-Tools.so => /usr/lib/libSPIRV-Tools.so (0x00007f2d728aa000)
        libxcb.so.1 => /usr/lib/libxcb.so.1 (0x00007f2d73a21000)
        libstdc++.so.6 => /home/me/gcc-env/lib64/libstdc++.so.6 (0x00007f2d72200000)
        libm.so.6 => /usr/lib/libm.so.6 (0x00007f2d73939000)
        libgcc_s.so.1 => /home/me/gcc-env/lib64/libgcc_s.so.1 (0x00007f2d73919000)
        libc.so.6 => /usr/lib/libc.so.6 (0x00007f2d72019000)
        libXau.so.6 => /usr/lib/libXau.so.6 (0x00007f2d73912000)
        libXdmcp.so.6 => /usr/lib/libXdmcp.so.6 (0x00007f2d7390a000)
        /lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007f2d73b0e000)
```

I hope you can see from this analysis what goes wrong.

**Comment by robertosfield** [comment]
Thanks for the details.  Attaching the CMakeCache.txt file for the VSG and vsgExamples is still helpful, so for the *fifth* time please can you attach it.  Is there a reason why you can't?  Others are able to attach on first request.

The ldd suggests that you have a mixture of GLSLang releated libs spread across your system.  I surprised that the VSG isn't just picking them all up from one location.  Could it be that your local gcc-env install has a different set of GLSLang related libs than the usr/lib ones, do it can find all the required libs from one location so falls back to others.

When you say "ignores my prefix" do you mean "-DCMAKE_INSTALL_PREFIX=${HOME}/gcc-env" ?

Could you let me know the versions of GLSLang in /usr/lib vs ${HOME}/gcc-env and the list of GLSLang related libs in each instance.

**Comment by intractabilis** [comment]
> Is there a reason why you can't?

First, there was a miscommunication. Now, after my analysis, I am 100% sure you have all the information needed. The problem is clear as the sky on a sunny day. You hacked CMake's usual process of finding libraries, and thus is the problem. But, anyway, attaching the files, as you requested.
[CMakeCache.tar.gz](https://github.com/vsg-dev/VulkanSceneGraph/files/10338690/CMakeCache.tar.gz)

> The ldd suggests that you have a mixture of GLSLang releated libs spread across your system.

I don't.

>  I surprised that the VSG isn't just picking them all up from one location.

It shouldn't.

> Could it be that your local gcc-env install has a different set of GLSLang related libs than the usr/lib ones

Yes. As I said before, I build GLSLang libraries and install them to my prefix.

> When you say "ignores my prefix" do you mean "-DCMAKE_INSTALL_PREFIX=${HOME}/gcc-env" ?

Yes. `CMAKE_INSTALL_PREFIX` is added to `CMAKE_SYSTEM_PREFIX_PATH`, [here](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html):

> The installation prefix is also added to [CMAKE_SYSTEM_PREFIX_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_PREFIX_PATH.html#variable:CMAKE_SYSTEM_PREFIX_PATH) so that [find_package()](https://cmake.org/cmake/help/latest/command/find_package.html#command:find_package), [find_program()](https://cmake.org/cmake/help/latest/command/find_program.html#command:find_program), [find_library()](https://cmake.org/cmake/help/latest/command/find_library.html#command:find_library), [find_path()](https://cmake.org/cmake/help/latest/command/find_path.html#command:find_path), and [find_file()](https://cmake.org/cmake/help/latest/command/find_file.html#command:find_file) will search the prefix for other software.

> Could you let me know the versions of GLSLang in /usr/lib vs ${HOME}/gcc-env and the list of GLSLang related libs in each instance.

It's irrelevant to the bug, but
```
$ yay -Qs glslang
local/glslang 11.11.0-1
    OpenGL and OpenGL ES shader front end and validator
$ grep -RP 'set\(PACKAGE_VERSION "\d' ${HOME}/gcc-env/lib/cmake/glslang
/home/user/gcc-env/lib/cmake/glslang/glslang-config-version.cmake:set(PACKAGE_VERSION "11.13.0")
 ```

**Comment by robertosfield** [comment]
Thanks for the CMakeCache.txt files.  It's end of day here now so I'll do a proper review tomorrow.

BTW, you contradict yourself, you can't say you don't have GLSLang releated libs spread across your system and then show that you have them in local and and gcc-env.  This is duplicity is crucial as it looks to be the what is causing the problem. Have a look through your ldd line.

As for the VSG provide it's own find scheme for GLSLang, this was forced upon us because older versions go GLSLang have no such support and neither does CMake, so we've had to create our own to handle this case.

**Comment by intractabilis** [comment]
I am having an impression you are trying to convince me there is something wrong with my environment. There is not. If you want my help in fixing this bug, let's not waste our energy on this, or find a contradiction in my words. What causing the problem is that you link with system libraries, instead of respecting the specified prefix. There is nothing unusual in my environment, it's a common development scenario: you, in order not to mess with the system packages, create a separate environment with your own toolset and compile your dependencies to this prefix with your toolset. It always works, unless the package author is doing some hacks and makes wrong assumptions. I've been doing this for years with hundreds of various libraries, there are only a couple that are causing problems.

> As for the VSG provide it's own find scheme for GLSLang, this was forced upon us because older versions go GLSLang have no such support and neither does CMake, so we've had to create our own to handle this case.

I understand this. However, you have broken the normal scenario to accommodate for weird cases. Why not check if the libraries you are looking for have CMake configuration files first, and only if not, fallback to your own search? Something like this:
```
find_package(glslang CONFIG COMPONENTS
    glslang
)
if (NOT glslang_FOUND)
    message(STATUS "glslang is not found")
else()
    message(STATUS "glslang is found")
endif()
```

**Comment by intractabilis** [comment]
Btw, CMake configuration modules do much more than just finding a library, they also check for dependencies, they set preprocessor macros that the authors expect their header files must be compiled with. For example, in case of the aforementioned glslang, CMake configuration file sets up `GLSLANG_IS_SHARED_LIBRARY` preprocessor variable. You simply ignore all of this.

**Comment by robertosfield** [comment]
I've been an open source project lead since 2000, have had to support thousands of developers and investigate thousands of issues.  Occasionally you get oddities that a small number of folks come across because of something they do slightly differently than everyone else that came before them.  

In the 4 years of the VSG you are the first to come across this specific issue, which is why right away I was asking for details about your specific usage case when why I was asking about CMakeCache.txt files.  In my experience it's these things that cut to the chase the quickest.  For some reason you've made it difficult to short cut things, most folks reporting issues are easier to work with as they just answer the questions and not try to spin things in one way or another.

**Comment by robertosfield** [comment]
I've been experimenting with a local build of glslang and found that vsg_glslang.cmake reports that glslang is not available because it's also checking for SPIRV-Tools and  SPIRV-Tools-opt which haven't been found as I haven't independently build and install SPIRV-Tools.  

Since the beginning of the VSG project glslang has kept evolving and moving the goal posts, changes w.r.t to SPIRV-Tools looks to another complication.  I am still experimenting so don't have any plan yet.  Unfortunately changing things has a big risk of breaking compatibility across platforms and with older systems so we'll need to be careful.

**Comment by intractabilis** [comment]
> For some reason you've made it difficult to short cut things

I have 35 years of experience. Since I have diagnosed where the problem is, the best shortcut could have been in trusting me in what I have found, instead of trying to spin the conversation your way. This is fine, if communication with me is so difficult for you, I am leaving the party.

**Comment by robertosfield** [comment]
On Wed, 4 Jan 2023 at 16:24, intractabilis ***@***.***> wrote:

> For some reason you've made it difficult to short cut things
>
> I have 35 years of experience. Since I have diagnosed where the problem is,
>
No, you awkwardly described a problem requiring lots of follow up questions
to clarify things, and haven't diagnosed anything, you only think you have.

I'm still investigating as to what the underlying cause and solution will
need to be.

**Comment by robertosfield** [comment]
I've rebuilt glslang and SPIRV-Tools as static and shared, installing to my /home/robert/Install direcotry on different tests and each combination works works.  The locally installed version of libglslang.so is correctly picked up.  I'm using -DCMAKE_INSTALL_PREFIX=~/Install in the build of glslang , SPIRV-Tools and all the VSG projects.

$ ldd ../vsgExamples/bin/vsgviewer | grep libgl
        libglslang.so.11 => /home/robert/Install/lib/libglslang.so.11 (0x00007ff9b4029000)

This is testing on Kubuntu 22.04 and with VSG master, and glslang and SPIRV-Tool master.

**Comment by robertosfield** [comment]
I have tried installing glslang using apt-get so it's located in the /usr/include + /usr/lib directory.  When I run cmake with -DCMAKE_INSTALL_PREFIX=~/Install it finds the /usr/ version of glslang rather my local one installed in ~/Install.

I have reviewed the relevant CMake docs and the relevant var that defines this default behavior is 

https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_PREFIX_PATH.html

When I print out ${CMAKE_SYSTEM_PREFIX_PATH} I get:

/usr/local/usr//usr/home/robert/Install/usr/X11R6/usr/pkg/opt

Which is strung together because the standard message() function in CMake doesn't output spacing for lists, so this is more readability:

    /usr/local
    /usr/
    /usr
    /home/robert/Install
   /usr/X11R6/usr/pkg/opt

Note CMAKE_I(NSTALL_PREFIX set path comes after the standard system diretories.  This looks to be standard for CMake and what find_path, find_library etc. all use by default.

CMake docs specify that if users want to override this then they should set CMAKE_PREFIX_PATH:

https://cmake.org/cmake/help/latest/variable/CMAKE_PREFIX_PATH.html#variable:CMAKE_PREFIX_PATH

Adding the following to my cmake invocation allows VSG etc. to pick up on my local install glslang in ~/Install:

    -DCMAKE_PREFIX_PATH=~/Install;~/usr/local;~/usr/

From the best I can work out the problem the user has was misunderstand of CMake's default behaviors and how to modify them, and the runtime issue was likely due to a building the VSG against one version of glslang but the runtime picking up on another at runtime due inconsistent paths for building and and runtume with LD_LIBRARY_PATH.

I'm closing this Issue as the VSG looks to behavior as CMake intends w,r.t precedence of default search paths.  I don't know whether this search precedence issues is something we need to document explicitly rather than just rely upon CMake docs so I'll ponder on this.
